### PR TITLE
[ORCT-132] CSV export

### PR DESCRIPTION
**Pull request by Ehevi**
### I have a [JIRA ticket](https://alice.its.cern.ch/jira/projects/ORCT/issues/ORCT-132?filter=allopenissues)
- [x] branch and/or PR name(s) include(s) JIRA ID
- [ ] issue has "Fix version" assigned
- [ ] issue "Status" is set to "In review"
- [x] PR labels are selected

### Notable changes for users:
- css export downloads all the items on page by default (no items selected)

### Notable changes for developers:
-

### CODE REVIEW

**Review by xsalonx — CHANGES_REQUESTED** [review]

File: app/public/utils/dataProcessing/dataProcessingUtils.js [hunk 1]
```diff
@@ -41,3 +41,16 @@ export const rowDisplayStyle = (isMarked, shouldHideMarkedRecords) => isMarked
...
+export const getReadableFileSizeString = (fileSizeInBytes) => {
+    let i = -1;
+    const byteUnits = ['kB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
+    do {
+        fileSizeInBytes /= 1024;
+        i++;
+    } while (fileSizeInBytes > 1024);
+
+    return `${Math.max(fileSizeInBytes, 0.1).toFixed(1)} ${byteUnits[i]}`;
+};
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataProcessing/dataProcessingUtils.js:54]: You are the author of this function, I just moved it to a different file <3

File: app/public/utils/dataExport/export.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+export const prepareCSVContent = (content) => {
+    const header = Object.keys(content[0]);
+    const csv = content.map((row) => header.map((fieldName) => JSON.stringify(row[fieldName], replacer)));
+    csv.unshift(header.join(','));
+    return csv.join('\r\n');
+};
+
+export const replacer = (_key, value) => value === null ? '' : value;
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataExport/export.js:66]: I have no idea how to explain this, but it doesn't work when I write `value ?? ''`;

Screenshot 2023-08-24 at 15 56 55

File: app/public/utils/dataExport/export.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * @param {String} fileName The name of the file including the output format.
+ * @param {String} contentType The content type of the file.
+ * @return {void}
+ */
+const downloadFile = (content, fileName, contentType) => {
+    const a = document.createElement('a');
+    const file = new Blob([content], { type: contentType });
+    a.href = URL.createObjectURL(file);
+    a.download = fileName;
+    a.click();
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataExport/export.js:32]: done

File: app/public/utils/dataProcessing/dataProcessingUtils.js [hunk 4]
```diff
@@ -41,3 +41,16 @@ export const rowDisplayStyle = (isMarked, shouldHideMarkedRecords) => isMarked
...
+
+export const capitalizeFirstLetter = (text) => text.charAt(0).toUpperCase() + text.slice(1);
```
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataProcessing/dataProcessingUtils.js]: We have actually already had this kind of function (pageTitle). I switched to using that one now.

[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataProcessing/dataProcessingUtils.js]: But I'll keep in mind the idea about organising it in a more structured way.

File: app/public/utils/dataProcessing/dataProcessingUtils.js [hunk 5]
```diff
@@ -41,3 +41,32 @@ export const rowDisplayStyle = (isMarked, shouldHideMarkedRecords) => isMarked
...
+        case pageNames.runsPerPeriod: return 'Runs per period';
+        case pageNames.runsPerDataPass: return 'Runs per data pass';
+        case pageNames.dataPasses: return 'Data passes per period';
+        case pageNames.mc: return 'Monte Carlo';
+        case pageNames.flags: return 'Quality flags';
+        case pageNames.anchoragePerDatapass: return 'Anchorage per data pass';
+        case pageNames.anchoredPerMC: return 'Anchored per MC';
+        default: return page;
+    }
+};
```
Comment (xsalonx) [inline app/public/utils/dataProcessing/dataProcessingUtils.js:72]: `pageTitle` doesn't belong to context of 'data processing utils'. I think it should be moved to other file

File: app/public/views/periods/overview/dataExport.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        const exportFormatSelection = document.getElementById(exportFormatSelectId);
+        const selectedFormat = exportFormatSelection.options[exportFormatSelection.selectedIndex].value;
+        switch (selectedFormat) {
+            case exportFormats.csv:
+                return selectedFormat;
+            case exportFormats.json:
+                return selectedFormat;
+            default:
+                return selectedFormat.csv;
+        }
```
Comment (xsalonx) [inline app/public/views/periods/overview/dataExport.js:107]: If it is not strictly necessary for those lambdas to be inside of main function body, they should be moved outside

File: test/public/utils/dataExport.test.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            ];
+
+            const preparedData = 'name,year,beamType,avgEnergy,distinctEnergies\r\n' +
+            `"LHC23zt",2023,"p-p",13595,[6797.04,6797.52,6797.64,6798]\r\n` +
+            `"LHC23zs",2023,"p-p",13596,[6797.4,6797.52,6797.64,6797.76,6797.88,6798,6798.24,6799.2]`;
+            
+            it('should not return null', () => {
+                assert(prepareCSVContent(rawData) !== null);
+            });
+
```
Comment (xsalonx) [inline test/public/utils/dataExport.test.js:69]: With regard of following one, that test in redundant

**Review by xsalonx — CHANGES_REQUESTED** [review]

File: app/public/utils/dataExport/export.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ehevi) [reply app/public/utils/dataExport/export.js]: Ok, that works.

File: app/public/views/periods/overview/dataExport.js (hunk 6 above)
Comment (Ehevi) [reply app/public/views/periods/overview/dataExport.js]: Fine, fixed that as well.

File: app/public/utils/dataProcessing/dataProcessingUtils.js (hunk 5 above)
Comment (Ehevi) [reply app/public/utils/dataProcessing/dataProcessingUtils.js]: I've made it a separate component.

**Review by xsalonx — APPROVED** [review]

**Pull request merged**